driver-core/src/main/com/mongodb/connection/DefaultServer.java

```diff
@@ -187,4 +220,44 @@ public void stateChanged(final ChangeEvent<ServerDescription> event) {
             serverListener.serverDescriptionChanged(new ServerDescriptionChangedEvent(serverId, description, oldDescription));
         }
     }
+
+    private final class ClusterClockAdvancingSessionContext implements SessionContext {
+
+        private SessionContext wrapped;
+
+        private ClusterClockAdvancingSessionContext(final SessionContext wrapped) {
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public boolean hasSession() {
+            return wrapped.hasSession();
+        }
+
+        @Override
+        public BsonDocument getSessionId() {
+            return wrapped.getSessionId();
+        }
+
+        @Override
+        public long advanceTransactionNumber() {
+            return wrapped.advanceTransactionNumber();
+        }
+
+        @Override
+        public void advanceOperationTime(final BsonTimestamp operationTime) {
+            wrapped.advanceOperationTime(operationTime);
+        }
+
+        @Override
+        public BsonDocument getClusterTime() {
+            return clusterClock.greaterOf(wrapped.getClusterTime());
```

Comment:
This implements the spec requirement that the cluster time sent with the command is the greater of the one associated with the session and the global cluster time.

Reply:
Ack.